core/contextmenu_registry.js

```diff
@@ -0,0 +1,172 @@
+ *    workspace: (Blockly.WorkspaceSvg|undefined),
+ * }}
+ */
+Blockly.ContextMenuRegistry.Scope;
+
+/**
+ * A menu item as entered in the registry.
+ * @typedef {{
+ *    callback: function(!Blockly.ContextMenuRegistry.Scope),
+ *    scopeType: !Blockly.ContextMenuRegistry.ScopeType,
+ *    displayText: ((function(!Blockly.ContextMenuRegistry.Scope):string)|string),
+ *    preconditionFn: function(!Blockly.ContextMenuRegistry.Scope):string,
+ *    weight: number,
+ *    id: string,
+ * }}
+*/
+Blockly.ContextMenuRegistry.RegistryItem;
+
+/**
+ * A menu item as presented to contextmenu.js.
+ * @typedef {{
+ *    text: string,
+ *    enabled: boolean,
+ *    callback: function(!Blockly.ContextMenuRegistry.Scope),
+ *    scope: !Blockly.ContextMenuRegistry.Scope,
+ *    weight: number,
+ * }}
+ */
+Blockly.ContextMenuRegistry.ContextMenuOption;
+
+/**
+ * Singleton instance of this class. All interactions with this class should be done on this object.
+ * @type {?Blockly.ContextMenuRegistry}
+ * @public
+ */
+Blockly.ContextMenuRegistry.registry = null;
+
+/**
+ * Registers a RegistryItem.
+ * @param {!Blockly.ContextMenuRegistry.RegistryItem} item Context menu item to register.
+ * @throws {Error} if an item with the given id already exists.
+ * @public
+ */
+Blockly.ContextMenuRegistry.prototype.register = function(item) {
+  if (this.registry_[item.id]) {
+    throw Error('Menu item with id "' + item.id + '" is already registered.');
+  }
+  this.registry_[item.id] = item;
+};
+
+/**
+ * Unregisters a RegistryItem with the given id.
+ * @param {string} id The id of the RegistryItem to remove.
+ * @throws {Error} if an item with the given id does not exist.
+ * @public
+ */
+Blockly.ContextMenuRegistry.prototype.unregister = function(id) {
+  if (this.registry_[id]) {
+    delete this.registry_[id];
+  } else {
+    throw new Error('Menu item with id "' + id + '" not found.');
+  }
+};
+
+/**
+ * @param {string} id The id of the RegistryItem to get.
+ * @returns {?Blockly.ContextMenuRegistry.RegistryItem} RegistryItem or null if not found
+ * @public
+ */
+Blockly.ContextMenuRegistry.prototype.getItem = function(id) {
+  if (this.registry_[id]) {
+    return this.registry_[id];
+  }
+  return null;
+};
+
+/**
+ * Gets the valid context menu options for the given scope type (e.g. block or workspace) and scope.
+ * Blocks are only shown if the preconditionFn shows they should not be hidden.
+ * @param {!Blockly.ContextMenuRegistry.ScopeType} scopeType Type of scope where menu should be
+ *     shown (e.g. on a block or on a workspace)
+ * @param {!Blockly.ContextMenuRegistry.Scope} scope Current scope of context menu
+ *     (i.e., the exact workspace or block being clicked on)
+ * @returns {!Array.<!Blockly.ContextMenuRegistry.ContextMenuOption>} the list of ContextMenuOptions
+ * @public
+ */
+Blockly.ContextMenuRegistry.prototype.getContextMenuOptions = function(scopeType, scope) {
+  var items = Object.keys(this.registry_);
```

Comment:
No need for the `items` var, can simply do `Object.keys(this.registry_).forEach...`

Reply:
Done